patches/server/0964-Support-registry-mod-API-with-GameEvent.patch

```diff
@@ -0,0 +1,132 @@
+ 
+ import com.google.common.collect.ImmutableList;
+ import io.papermc.paper.registry.entry.RegistryEntry;
++import io.papermc.paper.world.PaperGameEvent;
+ import io.papermc.paper.world.structure.ConfiguredStructure;
+ import io.papermc.paper.world.structure.PaperConfiguredStructure;
+ import java.util.IdentityHashMap;
+@@ -20,6 +21,7 @@ import org.checkerframework.checker.nullness.qual.Nullable;
+ 
+ import static io.papermc.paper.registry.entry.RegistryEntry.immutable;
+ import static io.papermc.paper.registry.entry.RegistryEntry.immutableBuiltIn;
++import static io.papermc.paper.registry.entry.RegistryEntry.writableBuiltIn;
+ 
+ public final class PaperRegistries {
+ 
+@@ -32,6 +34,7 @@ public final class PaperRegistries {
+     static {
+         REGISTRY_ENTRIES = ImmutableList.<RegistryEntry<?, ?>>builder()
+             // built-ins
++            .add(writableBuiltIn(RegistryKey.GAME_EVENT, Registries.GAME_EVENT, PaperGameEvent::new, io.papermc.paper.world.PaperGameEvent.Builder::new))
+             .add(immutableBuiltIn(RegistryKey.STRUCTURE_TYPE, Registries.STRUCTURE_TYPE, CraftStructureType::new))
+ 
+             // data-drivens
+diff --git a/src/main/java/io/papermc/paper/world/PaperGameEvent.java b/src/main/java/io/papermc/paper/world/PaperGameEvent.java
+new file mode 100644
+index 0000000000000000000000000000000000000000..0995cd6008486c784072a21f1990c935a801f5bb
+--- /dev/null
++++ b/src/main/java/io/papermc/paper/world/PaperGameEvent.java
+@@ -0,0 +1,59 @@
++package io.papermc.paper.world;
++
++import io.papermc.paper.registry.PaperRegistryBuilder;
++import org.bukkit.GameEvent;
++import org.bukkit.NamespacedKey;
++import org.checkerframework.checker.nullness.qual.NonNull;
++import org.checkerframework.checker.nullness.qual.Nullable;
++import org.checkerframework.framework.qual.DefaultQualifier;
++
++@DefaultQualifier(NonNull.class)
++public final class PaperGameEvent extends GameEvent {
++
++    private final net.minecraft.world.level.gameevent.GameEvent nms;
++
++    public PaperGameEvent(final NamespacedKey key, final net.minecraft.world.level.gameevent.GameEvent nms) {
++        super(key);
++        this.nms = nms;
++    }
++
++    @Override
++    public int getRange() {
++        return this.nms.getNotificationRadius();
++    }
++
++    public static final class Builder implements GameEvent.Builder, PaperRegistryBuilder<net.minecraft.world.level.gameevent.GameEvent, GameEvent> {
++
++        private final NamespacedKey key;
++        private int range = 16;
++
++        public Builder(NamespacedKey key, net.minecraft.world.level.gameevent.@Nullable GameEvent nms) {
++            this.key = key;
++            if (nms != null) {
++                this.range = nms.getNotificationRadius();
++            }
++        }
++
++        @Override
++        public NamespacedKey getKey() {
++            return this.key;
++        }
++
++        @Override
++        public int range() {
```

Comment:
Is there a way to do that without duplicating all the getter into the builder, this can become really complex on some other objects like Biome or DimensionType. I was thinking of something like that to have only the setters on the builder and then the getter on the final object:
```Java
 public void beforeRegister(final RegistryAdditionEvent<GameEvent, GameEvent.Builder> event) {
            if (event(.getOldObject()).key().equals(TO_MODIFY)) {
                event.builder().range(event.getOldObject().getRange() * 2);
            }
        }
```

Reply:
I don't think I'm a fan of that. I certainly don't like actually constructing the API type here. Sometimes not even the NMS type is constructed yet, and so that would require making API type impls not use nms types. 

I did consider having the builder also extend the api type, but I don't really like that either. It just feels too limiting if there are then methods added to the API type that aren't simple getters.